crates/prune/src/pruner.rs

```diff
@@ -196,133 +314,141 @@ impl<DB: Database> Pruner<DB> {
+            // Checkpoint exists, prune from the next transaction after the highest pruned one
+            .and_then(|checkpoint| match checkpoint.tx_number {
+                Some(tx_number) => Some(tx_number + 1),
+                _ => {
+                    error!(target: "pruner", %prune_part, ?checkpoint, "Expected transaction number in prune checkpoint, found None");
+                    None
+                },
+            })
             // No checkpoint exists, prune from genesis
             .unwrap_or(0);
 
-        // Get first transaction
-        let from_tx_num =
-            provider.block_body_indices(from_block_number)?.map(|body| body.first_tx_num);
-        // If no block body index is found, the DB is either corrupted or we've already pruned up to
-        // the latest block, so there's no thing to prune now.
-        let Some(from_tx_num) = from_tx_num else { return Ok(None) };
-
-        let to_tx_num = match provider.block_body_indices(to_block)? {
+        let to_tx_number = match provider.block_body_indices(to_block)? {
             Some(body) => body,
             None => return Ok(None),
         }
         .last_tx_num();
 
-        Ok(Some(from_tx_num..=to_tx_num))
+        Ok(Some(from_tx_number..=to_tx_number))
     }
 
-    /// Prune receipts up to the provided block, inclusive.
+    /// Prune receipts up to the provided block, inclusive, respecting the batch size.
     #[instrument(level = "trace", skip(self, provider), target = "pruner")]
     fn prune_receipts(
         &self,
         provider: &DatabaseProviderRW<'_, DB>,
         to_block: BlockNumber,
         prune_mode: PruneMode,
     ) -> PrunerResult {
-        let range = match self.get_next_tx_num_range_from_checkpoint(
+        let tx_range = match self.get_next_tx_num_range_from_checkpoint(
             provider,
             PrunePart::Receipts,
             to_block,
         )? {
             Some(range) => range,
             None => {
                 trace!(target: "pruner", "No receipts to prune");
-                return Ok(())
+                return Ok(true)
             }
         };
-        let total = range.clone().count();
+        let tx_range_end = *tx_range.end();
 
-        provider.prune_table_with_iterator_in_batches::<tables::Receipts>(
-            range,
+        let mut last_pruned_transaction = tx_range_end;
+        let (deleted, done) = provider.prune_table_with_range::<tables::Receipts>(
+            tx_range,
             self.batch_sizes.receipts,
-            |rows| {
-                trace!(
-                    target: "pruner",
-                    %rows,
-                    progress = format!("{:.1}%", 100.0 * rows as f64 / total as f64),
-                    "Pruned receipts"
-                );
-            },
+            |row| last_pruned_transaction = row.0,
         )?;
+        trace!(target: "pruner", %deleted, %done, "Pruned receipts");
+
+        let last_pruned_block = provider
```

Comment:
is this necessary ? doesn't seem to be used elsewhere for receipt pruning

Reply:
It will be used to restrict the changes of pruning configuration. For example, if user first had `receipts = { before = 1000 }`, we pruned all the receipts before block `1000`, and then configuration changed to `receipts = { before = 100 }`, so we need to backfill the receipts in range `100..=1000` which is not possible.

Also, it's just for the sake of expecting the same behaviour from all prune parts.